curp/src/keybased_channel.rs

```diff
@@ -118,31 +220,26 @@ enum MessageOrListener<M> {
     Listen(EventListener),
 }
 
-impl<K: Eq + Hash + Clone, M> KeybasedChannelReceiver<K, M> {
+impl<K: Eq + Hash + Clone + ConflictCheck, M> KeybasedChannelReceiver<K, M> {
     /// Get a message from the channel or a waiting listener
     #[allow(clippy::expect_used, clippy::unwrap_in_result)]
-    fn get_msg_or_listener(&self) -> MessageOrListener<M> {
+    fn get_msg_or_listener(&mut self) -> MessageOrListener<KeysMessage<K, M>> {
+        if let Some(buf_msg) = self.buf.pop_front() {
+            return MessageOrListener::Message(buf_msg);
+        }
+
         let mut channel = self.channel.lock();
-        let mut local_listener: Option<EventListener> = None;
-        loop {
-            match channel.try_recv() {
-                Some(msg) => return MessageOrListener::Message(msg),
-                #[allow(clippy::match_ref_pats)] // don't move listener out
-                None => match &local_listener {
-                    &Some(_) => {
-                        break;
-                    }
-                    &None => local_listener = Some(channel.new_msg_event.listen()),
-                },
-            }
+        self.buf.extend(channel.pop_all_msg().into_iter());
```

Comment:
possible to use mem::swap()?

Reply:
it's just one move fromone place to another, mem::swap is not more efficient in this case I think.